### PR TITLE
Kmc surrogate recipe

### PR DESCRIPTION
**Pull request by CharlelieLrt**
# PhysicsNeMo Pull Request

## Description

Adds a new example under `examples/kinetic_monte_carlo`: a probabilistic, autoregressive surrogate for Kinetic Monte Carlo (KMC) solvers. The model (`ParticleGeoTransolver`) emulates a KMC event stream one event at a time, predicting the next event (the new particle's features and the inter-event delay) from the current particle population, an optional background mesh, and the simulation time. Because the prediction is probabilistic, independent rollouts from the same initial state form an ensemble that quantifies the uncertainty of the underlying stochastic process. The recipe is general purpose (agnostic to what a "particle" represents) and ships Hydra-configured scripts for dataset statistics, training, and ensemble generation, plus a README documenting the datapipe format. It is self-contained and adds no new framework dependencies.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Dependencies

None for the core library. The example declares its own extras in `examples/kinetic_monte_carlo/requirements.txt`.

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: examples/kinetic_monte_carlo/train.py [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+        current_samples_trained = int(metadata["current_samples_trained"])
+        rank_zero.info(f"Resuming at samples trained: {current_samples_trained}")
+    total_batch_size = cfg.training.batch_size_per_gpu * dist.world_size
+    rank_zero.info(
+        f"Batch size: per-GPU={cfg.training.batch_size_per_gpu} | "
+        f"total (across {dist.world_size} ranks)={total_batch_size}"
+    )
+    sampler_start_idx = current_samples_trained
+
+    train_loader = ParticlesDataPipe(
```
Comment (ktangsali) [inline examples/kinetic_monte_carlo/train.py:200]: No val / test dataloader ?

File: examples/kinetic_monte_carlo/README.md
Comment (ktangsali) [inline examples/kinetic_monte_carlo/README.md:1]: We should link this README to the overall README present in the examples directory too. Additionally, don't forget to update the docs repo once this is merged - I think it would be good to highlight this!

File: examples/kinetic_monte_carlo/README.md [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+command line. The workflow is:
+
+1. [Dataset format](#dataset-format)
+2. [Compute dataset statistics](#compute-dataset-statistics)
+3. [Training](#training)
+4. [Generation and rollout](#generation-and-rollout)
+
+## Dataset format
+
+This example does **not** ship a preprocessing pipeline: you are expected to
```
Comment (ktangsali) [inline examples/kinetic_monte_carlo/README.md:104]: Would be nice to think of a synthetic dataset generator or possibly shipping a toy dataset. Not a blocker for this, but it will increase the usability of this :)

File: examples/kinetic_monte_carlo/compute_stats.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    delay_stats = Welford()
+    log_delay_stats = Welford()
+    particle_feature_stats = {name: Welford() for name in particle_feature_names}
+    mesh_feature_stats = {name: Welford() for name in mesh_feature_names}
+
+    for i, (batch, _meta) in enumerate(loader):
+        mask = batch["particle_state"] > 0.5  # (B, S, K)
+
+        # Spatial coords: pool mesh and real-particle coordinates (same unit).
+        coord_stats.update(batch["mesh_coords"])  # (B, N, 3)
```
Comment (ktangsali) [inline examples/kinetic_monte_carlo/compute_stats.py:185]: This pools mesh and particle coordinates into the same accumulator, and Welford.update() flattens all dimensions, producing one scalar mean/std shared by x, y, and z. Is that intentional?

That works very well for isotropic geometries, but for anisotorpic ones, having them different might be better?

File: examples/kinetic_monte_carlo/README.md [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+solver (defects, adatoms, traps, nucleated grains, ...), but the code is
+agnostic to what a particle represents.
+
+At step $n$ the surrogate maps the current state to the **next event** the
+solver would produce. Its inputs are:
+
+- the **particle population**: each particle's coordinates $(x, y, z)$, scalar
+  features $X_p$, and the delay since the previous event;
+- the current simulation **time** $t_n$;
+- *(optional)* a static background **mesh** describing the geometry the events
```
Comment (ktangsali) [inline examples/kinetic_monte_carlo/README.md]: The mesh is described as optional and this says the surrogate can run on the particle stream alone. However, `ParticlesDataset` currently requires a `maps/` directory and validates that every simulation has a `maps_<sim>.pth` file, and I do not see a configuration path that disables mesh inputs. Am I missing something?

File: examples/kinetic_monte_carlo/conf/config_generate.yaml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+# What to roll out. Each entry identifies a simulation by its geometry-group
+# directory name and its sim_id. Use ``sim_id: "all"`` to expand into every
+# simulation available for that geometry group.
+rollout:
+  # Number of independent trajectories generated per simulation. They share the
+  # same initial state and mesh but diverge through independent GMM draws.
+  num_ensemble: 8
+  simulations:
+    - geometry: "2nm"
```
Comment (ktangsali) [inline examples/kinetic_monte_carlo/conf/config_generate.yaml]: Maybe we call this something generic?

File: examples/kinetic_monte_carlo/compute_stats.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+        if (
+            not torch.distributed.is_available()
+            or not torch.distributed.is_initialized()
+        ):
+            return
+        world_size = torch.distributed.get_world_size()
+        if world_size < 2:
+            return
+
+        triples = torch.tensor([self.count, self.mean, self.M2], dtype=torch.float64)
```
Comment (ktangsali) [inline examples/kinetic_monte_carlo/compute_stats.py]: `DistributedManager` normally selects NCCL when CUDA is available, but `triples` and `gathered` are created on the CPU. NCCL collectives require CUDA tensors, so I think the multi-GPU `compute_stats.py` command documented in the README will fail here. Could we verify? Could these tensors be created on the distributed device?

File: examples/kinetic_monte_carlo/generate.py [hunk 7]
```diff
@@ -0,0 +1,630 @@
...
+            # Start the rollout from whatever particles the source data carries.
+            initial = {
+                "coords": coords[0].clone(),  # (K, 3)
+                "features": scalars[0].clone(),  # (K, P)
+                "delay": cast(Tensor, sample["delay"])[0].clone(),  # (K,)
+                "state": cast(Tensor, sample["particle_state"])[0].clone(),  # (K,)
+            }
+            initial_time = float(cast(Tensor, sample["time"])[0].item())
+            continue
+        # The most recently created particle lives at slot index (ts_id - 1).
```
Comment (ktangsali) [inline examples/kinetic_monte_carlo/generate.py]: The README allows sample_*_0 to contain a possibly non-empty initial state, and the code above preserves that population. In that case, the particle created at event ts_id is not necessarily at ts_id - 1; it should be offset by the initial particle count. Could this identify the newly added slot from consecutive state masks/counts, or should the data contract explicitly require an empty initial state?

File: docs/img/kmc_energy_rollout.gif
Comment (ktangsali) [inline docs/img/kmc_energy_rollout.gif:1]: Is this GIF low resolution because it was compressed to meet the repo demands? If so, can we tweak the compression parameters a bit to produce a variant where the text is visible a bit more better?

**Review by ktangsali — APPROVED** [review]
This looks great! Mostly some minor comments

File: examples/kinetic_monte_carlo/dataset/dataset.py [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+
+        Keys are ``"mesh_coords"`` of shape ``(N, 3)`` plus one ``(N,)``
+        tensor per entry of ``mesh_feature_names``.
+        """
+        key = (geometry, sim_id)
+        cached = self._mesh_cache.get(key)
+        if cached is not None:
+            return cached
+        path = self._maps_root / geometry / f"maps_{sim_id}.pth"
+        maps = torch.load(path, weights_only=False)
```
[the remark this answers is not in the document]
Comment (CharlelieLrt) [reply examples/kinetic_monte_carlo/dataset/dataset.py:273]: Thanks. I'm keeping `weights_only=False` here for now: this example runs on data the user preprocesses themselves (nothing is bundled or fetched from a third party), so the file provenance is under their control. Happy to switch to `weights_only=True` if we ever ship or download data from an external source.

File: examples/kinetic_monte_carlo/dataset/dataset.py [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+            for path in geometry_dir.glob("sample_*_*.pth"):
+                m = _SAMPLE_RE.fullmatch(path.name)
+                if m is None:
+                    continue
+                sim_id = int(m.group(1))
+                ts = int(m.group(2))
+                if ts > max_ts.get(sim_id, -1):
+                    max_ts[sim_id] = ts
+            for sim_id, n_events in max_ts.items():
+                self._per_sim_count[(geometry, sim_id)] = n_events
```
[the remark this answers is not in the document]
Comment (CharlelieLrt) [reply examples/kinetic_monte_carlo/dataset/dataset.py]: Good call. `_scan_samples` now collects every `ts` per simulation and raises a named error at construction if they don't form `range(0, max_ts + 1)`, so a gap is reported up front instead of as a bare `FileNotFoundError` mid-training. Fixed in 906b0f81.

File: examples/kinetic_monte_carlo/generate.py [hunk 10]
```diff
@@ -0,0 +1,630 @@
...
+    while not bool(done.all()):
+        iteration += 1
+        # Normalize the evolving granular state (mesh was normalized once above).
+        coords_n = _z_score(coords_raw, stats["coord_mean"], stats["coord_std"])
+        scalars_n = _z_score(scalars_raw, pf_mean, pf_std)
+        delay_n = _z_score(delay_raw, stats["delay_mean"], stats["delay_std"])
+        t_n = _z_score(t_raw, stats["delay_mean"], stats["delay_std"])
+
+        with torch.no_grad():
+            h_g = model(
```
[the remark this answers is not in the document]
Comment (CharlelieLrt) [reply examples/kinetic_monte_carlo/generate.py]: Agreed. Hoisted it to function scope with `@torch.no_grad()` on `_rollout_ensemble` and dropped the per-step `with`. 27eb7027.

File: examples/kinetic_monte_carlo/train.py (hunk 1 above)
Comment (CharlelieLrt) [reply examples/kinetic_monte_carlo/train.py:200]: Added an optional held-out split. The datapipe now takes a `phase` argument (`"train"`/`"test"`/`"all"`, default `"all"`), and a `dataset.train_test_split` config toggles it: `false` trains on everything under `data_dir`; `true` reads `train/` and `test/` subdirectories and evaluates a test loss every logging step, averaged and reported alongside the train loss. The layout is documented in the README dataset-format section. cf8cafc3.

File: examples/kinetic_monte_carlo/README.md
Comment (CharlelieLrt) [reply examples/kinetic_monte_carlo/README.md:1]: Moved it into its own Kinetic Monte Carlo section in `examples/README.md` (5a4ab3fd). And thanks for the docs-repo reminder, I'll add it there once this merges.

File: examples/kinetic_monte_carlo/README.md (hunk 2 above)
Comment (CharlelieLrt) [reply examples/kinetic_monte_carlo/README.md:104]: Agreed, a toy dataset (or a small synthetic generator) would help usability. I'll track it as a follow-up rather than fold it into this PR, since it isn't blocking.

File: examples/kinetic_monte_carlo/compute_stats.py (hunk 3 above)
Comment (CharlelieLrt) [reply examples/kinetic_monte_carlo/compute_stats.py:185]: Yes, intentional. A single shared coordinate stat (one mean/std across x, y, z and across particles and mesh) preserves the geometry's aspect ratios. Per-axis normalization would rescale the axes independently and distort them (a thin elongated slab would look closer to a cube), and aspect ratio is very sensitive here: small changes can move the predictions by orders of magnitude. Per-axis stats could be a future option for strongly anisotropic geometries, but the shared stat is deliberate.

File: examples/kinetic_monte_carlo/README.md (hunk 4 above)
Comment (CharlelieLrt) [reply examples/kinetic_monte_carlo/README.md]: You're right, the wording was misleading. As shipped the mesh is required: the datapipe expects a `maps/` file per simulation and the model forward consumes mesh tokens, with no config to disable it. I reworded the overview to state the mesh is a required input and flagged a mesh-free variant as a possible future extension. 1c115118.

File: examples/kinetic_monte_carlo/conf/config_generate.yaml (hunk 5 above)
Comment (CharlelieLrt) [reply examples/kinetic_monte_carlo/conf/config_generate.yaml]: Renamed it to a generic `my_geometry` placeholder, with a comment noting geometry-group names are user-defined. 820b9890.

File: examples/kinetic_monte_carlo/compute_stats.py (hunk 6 above)
Comment (CharlelieLrt) [reply examples/kinetic_monte_carlo/compute_stats.py]: Good catch, that would fail under NCCL. The Welford all-reduce built its tensors on CPU; they're now created on `DistributedManager().device` (and read back with `.tolist()`), so the documented multi-GPU command works. Fixed in 43441e47.

File: examples/kinetic_monte_carlo/generate.py (hunk 7 above)
Comment (CharlelieLrt) [reply examples/kinetic_monte_carlo/generate.py]: Made it robust to a non-empty initial state: the newly-created particle is now taken as the newest occupied slot, derived from the live count (`particle_state.sum() - 1`) rather than `ts_id - 1`. 1eec1afb.

File: docs/img/kmc_energy_rollout.gif
Comment (CharlelieLrt) [reply docs/img/kmc_energy_rollout.gif:1]: Regenerated it at higher resolution. The previous gif had been downscaled and heavily compressed to keep the file small, which is what softened the text. The new one is rendered at higher DPI and only lightly compressed (1.3 MB, well under the repo limit), so the labels are legible; it also picks up the current wording. 2bb91975.

**Comment by CharlelieLrt** [comment]
/ok to test [1f8b97e](https://github.com/NVIDIA/physicsnemo/pull/1747/commits/1f8b97ed837d9a251eb24774a2e879052720a696)
